# Overview
Gemma (pronounced: _Jemma_) is database of curated and re-analyzed gene expression studies. 

## Key features:
- All data sets are manually curated and QC'd in a standardized pipeline.
- Support for a variety of expression technologies, including scRNA-seq, bulk RNA-seq, Affymetrix, Illumina and other oligonucleotide arrays, one channel and ratiometric cDNA arrays.
- Re-annotated microarray platforms at the sequence level, which allows for more consistent cross-platform comparisons.
- Re-processed raw sequencing data (single cell and bulk tissue RNA-seq).
- Re-analysis of data sets for differential expression
- GUI and programmatic access

## Explore topics: 

 - **[Terms and conditions](terms.md)**
 - **[Glossary](glossary.md)**
 - **Using the Gemma website**
 - **[Data curation](curation.md)**
 - **RNA-seq processing pipeline**
 - **[single cell RNA-seq](scrnaseq.md)**
 - **[Programmatic use of Gemma](programmatic.md)**


## Data sources
We are indebted to the many researchers who have made data publicly available. Lists of published papers that relate to the data included in Gemma are available [here (full list)](https://gemma.msl.ubc.ca/bibRef/showAllEeBibRefs.html) and [here (search)](https://gemma.msl.ubc.ca/bibRef/searchBibRefs.html).

If your data is in Gemma, and your paper is not listed, please let us know.

## Contact

If you find a problem or need help, you can file a new github issue, or contact us at [pavlab-support@msl.ubc.ca](mailto:pavlab-support@msl.ubc.ca).

## Credits

### Financial support

As of 2023, Gemma is primarily supported by a grant from NIMH, and additional support from NSERC and CFI, for which we are grateful!

### Citing

If you use any of Gemma tools or data for your research, please cite one of the following papers:

[Lim et al. Curation of over 10 000 transcriptomic studies to enable data reuse. Database, 2021](https://doi.org/10.1093/database/baab006)

[Zoubarev, A., et al., Gemma: A resource for the re-use, sharing and meta-analysis of expression profiling data. Bioinformatics, 2012.](http://dx.doi.org/doi:10.1093/bioinformatics/bts430)

**Project lead:**
[Paul Pavlidis, Ph.D.](http://pavlab.msl.ubc.ca/paul-pavlidis/)


Copyright Â© University of British Columbia
