/*
 * The Gemma project
 *
 * Copyright (c) 2007 University of British Columbia
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package ubic.gemma.core.analysis.sequence;

import ubic.gemma.model.genome.sequenceAnalysis.BlatResult;
import ubic.gemma.core.config.Settings;

/**
 * Used to convert BlatResult objects into PSL lines that can be displayed in the UCSC Genome Browser.
 *
 * @author pavlidis
 */
public class BlatResult2Psl {

    private static final int EXTRA_WINDOW = 1000;

    /**
     * See <a href='http://genome.ucsc.edu/goldenPath/help/customTrack.html'>golden path custom track help</a>.
     *
     * @param blatResult blat result
     * @return psl
     */
    private static String blatResult2Psl( BlatResult blatResult ) {
        StringBuilder buf = new StringBuilder();
        buf.append( blatResult.getMatches() ).append( " " );
        buf.append( blatResult.getMismatches() ).append( " " );
        buf.append( blatResult.getRepMatches() ).append( " " );
        buf.append( blatResult.getNs() ).append( " " );
        buf.append( blatResult.getQueryGapCount() ).append( " " );
        buf.append( blatResult.getQueryGapBases() ).append( " " );
        buf.append( blatResult.getTargetGapCount() ).append( " " );
        buf.append( blatResult.getTargetGapBases() ).append( " " );
        buf.append( blatResult.getStrand() ).append( " " );
        buf.append( "\"" ).append( blatResult.getQuerySequence().getName() ).append( "\" " );
        buf.append( blatResult.getQuerySequence().getLength() ).append( " " );
        buf.append( blatResult.getQueryStart() ).append( " " );
        buf.append( blatResult.getQueryEnd() ).append( " " );
        buf.append( "chr" ).append( blatResult.getTargetChromosome().getName() ).append( " " );
        if ( blatResult.getTargetChromosome() != null && blatResult.getTargetChromosome().getSequence() != null
                && blatResult.getTargetChromosome().getSequence().getLength() != null ) {
            buf.append( blatResult.getTargetChromosome().getSequence().getLength() ).append( " " );
        } else if ( blatResult.getTargetSequence() != null && blatResult.getTargetSequence().getLength() != null ) {
            buf.append( blatResult.getTargetSequence().getLength() ).append( " " );
        } else {
            buf.append( blatResult.getTargetEnd() + 1 ).append( " " );// seems okay as long as more than the target end.
        }
        buf.append( blatResult.getTargetStart() ).append( " " );
        buf.append( blatResult.getTargetEnd() ).append( " " );
        buf.append( blatResult.getBlockCount() ).append( " " );
        buf.append( blatResult.getBlockSizes() ).append( " " );
        buf.append( blatResult.getQueryStarts() ).append( " " );
        buf.append( blatResult.getTargetStarts() );
        buf.append( "\n" );
        return buf.toString();
    }

    /**
     * Creates text that can be displayed directly as a track in UCSC, using their hgTracks program.
     * <a href='http://genome.ucsc.edu/goldenPath/help/customTrack.html'>golden path custom track help</a>.
     *
     * @param blatResult blat result
     * @return psl track
     */
    public static String blatResult2PslTrack( BlatResult blatResult ) {

        return "## Generated by Gemma (" + Settings.getHostUrl() + ")\n" + "browser position " + "chr" + blatResult
                .getTargetChromosome().getName() + ":" + ( blatResult.getTargetStart() - EXTRA_WINDOW ) + "-" + (
                blatResult.getTargetEnd() + EXTRA_WINDOW ) + "\n" + "track name=\"" + blatResult.getQuerySequence()
                .getName() + "\"" + " description=\"Gemma BLAT alignment\" visibility=2 useScore=1\n" + blatResult2Psl(
                blatResult ) + "\n";
    }

}
