*** applicationContext-hibernate.bad.xml	Wed Mar  2 19:52:37 2005
--- applicationContext-hibernate.xml	Wed Mar  2 19:57:31 2005
***************
*** 23,33 ****
          <property name="dataSource"><ref local="dataSource"/></property>
          <property name="mappingResources">
              <list>
!                 <value>edu/columbia/gemma/expression/bioAssayData/DataVector.hbm.xml</value>
!                 <value>edu/columbia/gemma/common/auditAndSecurity/AuditEvent.hbm.xml</value>
!                 <value>edu/columbia/gemma/common/Identifscribable.hbm.xml</value>
!                 <value>edu/columbia/gemma/common/auditAndSecurity/AuditTrail.hbm.xml</value>
!                 <value>edu/columbia/gemma/common/description/OntologyEntry.hbm.xml</value>
              </list>
          </property>
          <property name="hibernateProperties">
--- 23,92 ----
          <property name="dataSource"><ref local="dataSource"/></property>
          <property name="mappingResources">
              <list>
!                <value>edu/columbia/gemma/common/description/Description.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/Person.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/User.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/bioAssayData/DesignElementDataVector.hbm.xml</value>
!         <value>edu/columbia/gemma/common/protocol/Protocol.hbm.xml</value>
!         <value>edu/columbia/gemma/common/bqs/BibliographicReference.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/Gene.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/OtherGeneCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/sequenceAnalysis/BlatResult.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/experiment/ExpressionExperimentSubSet.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/HumanGeneCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/common/measurement/Measurement.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/MouseGeneCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/common/measurement/Unit.hbm.xml</value>
!         <value>edu/columbia/gemma/common/protocol/Software.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/Taxon.hbm.xml</value>
!         <value>edu/columbia/gemma/analysis/GeneInvestigation.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/FileFormat.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/ResearchGroup.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/MappingCandidateGene.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/Security.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/GenomicROI.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/Contact.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/OtherProbeCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/OntologyEntry.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/Chromosome.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/HumanProbeCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/AuditTrail.hbm.xml</value>
!         <value>edu/columbia/gemma/association/GeneHomology.hbm.xml</value>
!         <value>edu/columbia/gemma/association/BlatAssociation.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/ExternalReference.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/designElement/DesignElement.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/experiment/ExperimentalFactor.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/File.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/RatGeneCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/analysis/Analysis.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/bioAssay/BioAssay.hbm.xml</value>
!         <value>edu/columbia/gemma/association/Gene2GOAssociation.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/GeneProduct.hbm.xml</value>
!         <value>edu/columbia/gemma/analysis/AnalysisResult.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/experiment/FactorValue.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/experiment/ExperimentalDesign.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/BibReferenceCandidateGene.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/AuditEvent.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/bioAssayData/BioAssayDataVector.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/DatabaseEntry.hbm.xml</value>
!         <value>edu/columbia/gemma/association/BlastAssociation.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/biosequence/BioSequence.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/SecurityGroup.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/experiment/ExpressionExperiment.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/CandidateGene.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/biomaterial/Treatment.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/MouseProbeCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/common/protocol/Parameter.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/sequenceAnalysis/BlastResult.hbm.xml</value>
!         <value>edu/columbia/gemma/association/coexpression/RatProbeCoExpression.hbm.xml</value>
!         <value>edu/columbia/gemma/sequence/gene/GeneAlias.hbm.xml</value>
!         <value>edu/columbia/gemma/common/description/ExternalDatabase.hbm.xml</value>
!         <value>edu/columbia/gemma/common/auditAndSecurity/Organization.hbm.xml</value>
!         <value>edu/columbia/gemma/common/quantitationtype/QuantitationType.hbm.xml</value>
!         <value>edu/columbia/gemma/association/ReferenceAssociation.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/arrayDesign/ArrayDesign.hbm.xml</value>
!         <value>edu/columbia/gemma/expression/biomaterial/BioMaterial.hbm.xml</value>
! 
              </list>
          </property>
          <property name="hibernateProperties">
